File: projects/agentic-automation-framework-python/agentic/agent/memory.py
import json
import re
from pathlib import Path

REPORTS = Path("reports")
MEM = REPORTS / "agent-memory.json"
SEM = REPORTS / "agent-memory-embeddings.json"


def _ensure():
    REPORTS.mkdir(exist_ok=True)


def load_memory():
    _ensure()
    if not MEM.exists():
        return {"failures": []}
    try:
        return json.loads(MEM.read_text())
    except Exception:
        return {"failures": []}


def record_failure(item: dict):
    m = load_memory()
    m["failures"].append(item)
    m["failures"] = m["failures"][-50:]
    MEM.write_text(json.dumps(m, indent=2))


def recent(limit=5):
    return load_memory().get("failures", [])[-limit:]


def _tokens(s: str):
    return set([t for t in re.sub(r"[^a-z0-9\s]", " ", (s or "").lower()).split() if len(t) > 2])


def _jaccard(a, b):
    inter = len(a.intersection(b))
    union = len(a.union(b)) or 1
    return inter / union


def relevant_lexical(goal: str, limit=5):
    g = _tokens(goal)
    scored = []
    for f in load_memory().get("failures", []):
        txt = f"{f.get('goal','')} {f.get('error','')}"
        s = _jaccard(g, _tokens(txt))
        if s > 0:
            x = dict(f)
            x["score"] = s
            scored.append(x)
    scored.sort(key=lambda x: x["score"], reverse=True)
    return scored[:limit]

File: projects/agentic-automation-framework/src/agent/memory.js
import fs from "fs";
import path from "path";

const REPORT_DIR = path.resolve("reports");
const MEMORY_FILE = path.join(REPORT_DIR, "agent-memory.json");

function ensure() {
  fs.mkdirSync(REPORT_DIR, { recursive: true });
}

export function loadMemory() {
  ensure();
  if (!fs.existsSync(MEMORY_FILE)) return { failures: [] };
  try {
    return JSON.parse(fs.readFileSync(MEMORY_FILE, "utf-8"));
  } catch {
    return { failures: [] };
  }
}

export function recordFailure({ goal, error, workflowFile, iteration }) {
  const mem = loadMemory();
  mem.failures.push({
    ts: new Date().toISOString(),
    goal,
    error,
    workflowFile,
    iteration
  });
  mem.failures = mem.failures.slice(-50);
  fs.writeFileSync(MEMORY_FILE, JSON.stringify(mem, null, 2));
}

export function recentFailureContext(limit = 5) {
  const mem = loadMemory();
  return (mem.failures || []).slice(-limit);
}

function tokenize(s) {
  return new Set(
    (s || "")
      .toLowerCase()
      .replace(/[^a-z0-9\s]/g, " ")
      .split(/\s+/)
      .filter((x) => x.length > 2)
  );
}

function jaccard(a, b) {
  const inter = new Set([...a].filter((x) => b.has(x))).size;
  const union = new Set([...a, ...b]).size || 1;
  return inter / union;
}

export function relevantFailureContext(goal, limit = 5) {
  const mem = loadMemory();
  const goalTokens = tokenize(goal);
  const scored = (mem.failures || []).map((f) => {
    const txt = `${f.goal || ""} ${f.error || ""}`;
    const score = jaccard(goalTokens, tokenize(txt));
    return { ...f, score };
  });

  return scored
    .filter((x) => x.score > 0)
    .sort((a, b) => b.score - a.score)
    .slice(0, limit);
}
